pkg/storage/repo/storage.go

```diff
@@ -0,0 +1,331 @@
+	if err != nil {
+		return nil, err
+	}
+	dir, err := os.MkdirTemp(config.GetStoragePath(), "bacalhau-repo")
+	if err != nil {
+		return nil, err
+	}
+	cm.RegisterCallback(func() error {
+		if err := os.RemoveAll(dir); err != nil {
+			return fmt.Errorf("unable to remove storage folder: %w", err)
+		}
+		return nil
+	})
+	storageHandler := &StorageProvider{
+		LocalDir:      dir,
+		EstuaryAPIKey: EstuaryAPIKey,
+		IPFSClient:    IPFSapiclient,
+		CloneClient:   c,
+	}
+	log.Debug().Msgf("URL download driver created with output dir: %s", dir)
+	return storageHandler, nil
+}
+
+func (sp *StorageProvider) IsInstalled(context.Context) (bool, error) {
+	err := checkGitLFS()
+	if err != nil {
+		return false, err
+	} else {
+		return true, nil
+	}
+}
+
+func (sp *StorageProvider) HasStorageLocally(context.Context, model.StorageSpec) (bool, error) {
+	return false, nil
+}
+
+// Could do a HEAD request and check Content-Length, but in some cases that's not guaranteed to be the real end file size
+func (sp *StorageProvider) GetVolumeSize(context.Context, model.StorageSpec) (uint64, error) {
+	return 0, nil
+}
+
+func (sp *StorageProvider) PrepareStorage(ctx context.Context, storageSpec model.StorageSpec) (storage.StorageVolume, error) {
+	_, span := system.GetTracer().Start(ctx, "pkg/storage/repo/repo.PrepareStorage")
+	defer span.End()
+	repoURL := storageSpec.Repo
+	_, err := IsValidGitRepoURL(repoURL)
+	// fmt.Printf("%+v", storageSpec)
+	if err != nil {
+		return storage.StorageVolume{}, err
+	}
+	// # create a tmp directory
+	outputPath, err := os.MkdirTemp(sp.LocalDir, "*")
+	fmt.Printf("Output Path: %s\n", outputPath)
+	if err != nil {
+		return storage.StorageVolume{}, err
+	}
+	cmd := exec.Command("git", "clone", repoURL, outputPath)
+	// The `Output` method executes the command and
+	// collects the output, returning its value
+	out, err := cmd.Output()
+	if err != nil {
+		// if there was any error, print it here
+		fmt.Println("could not run command: ", err)
+	}
+	// otherwise, print the output from running the command
+	fmt.Println("Output: ", string(out))
+	if err != nil {
+		panic(err)
+	}
+	filepath, err := url.Parse(repoURL)
+	filename := strings.Split(filepath.Path, ".")[0]
+	targetPath := "/inputs" + filename
+	if err != nil {
+		return storage.StorageVolume{}, err
+	}
+	CIDSpec, err := sp.Upload(ctx, outputPath)
+	// If estuary key exists then upload it to estuary
+	if err != nil {
+		return storage.StorageVolume{}, err
+	}
+	CID := CIDSpec.CID
+	// Update the KV store
+	SHA1HASH, _ := UrltoLatestCommitHash(repoURL)
+	spkey := sp.EstuaryAPIKey
+	envkey := os.Getenv("ESTUARY_API_KEY")
+	var key string
+	if spkey != "" && envkey == "" {
+		key = spkey
+	} else if envkey != "" && spkey == "" {
+		key = envkey
+	} else if envkey != "" && spkey != "" {
+		key = spkey
+	} else {
+		key = ""
+	}
+	if key != "" {
+		log1.Println("Pinning to Estuary...")
+		//nolint:govet,noctx
+		err := estuary.PinToIPFSViaEstuary(ctx, key, CID)
```

Comment:
why are we cloning the repo to IPFS? I thought we just needed to mount the repo locally as a job input. Maybe adding some description to the PR will help me understand the feature more

Reply:
Sorry I didn't add much description here but the point with saving it to IPFS is that subsequent mounts would be done from IPFS instead of cloning the repo every time a person wants to mount a git repo


There is a key value (Key=SHA1hash, Value=CID) store hosted at kv.bacalhau.org where we make queries against the SHA hash of a git repo and check it if the CID exists in the store if it does we mount the CID instead of the repo, if the CID isn't there we clone the repo and post its's value in the kv store